stan/math/opencl/cholesky_decompose.hpp

```diff
@@ -54,12 +52,12 @@ inline matrix_cl cholesky_decompose(matrix_cl& A) {
     matrix_cl L(A.rows(), A.cols());
     try {
       opencl_kernels::cholesky_decompose(cl::NDRange(A.rows()),
-                                         cl::NDRange(A.rows()), A.buffer(),
-                                         L.buffer(), A.rows());
+                                         cl::NDRange(A.rows()), A, L, A.rows());
     } catch (const cl::Error& e) {
       check_opencl_error("cholesky_decompose", e);
     }
-    return L;
+    A = L;
```

Comment:
So, this is sort of a weirdo I've been trying to munge over.

In the old version of this we would _both_ manipulate the input A and return it. Is that okay? My thought here was that if we have A as input/output the return should be void to make it clear that A is what is being manipulated. If that's a misguided thought on my part I'm cool with making this return A

Reply:
Ahh interesting! Sorry, didn't realize that - you are completely correct & this is a great correction. Is it worth changing the kernel to just modify A then?

Reply:
Pinging @rok-cesnovar, how hard would doing the kernel inplace be? I looked over magma's and got a bit confused. Looking over our code I think it would be a solid rewrite.

If Rok thinks it's not hard then that would be cool, else I think it's fine to leave this as is

Reply:
Yeah, should be pretty straightforward to replace this with something like:
```
DO  I = 1, N
		A(I,I) = SQRT (A(I, I))
		DO  J = I+1, N
			A(J,I) = A(J,I)/A(I,I)
		END DO
		DO  K=I+1,N
			DO J = K, N
				A(J,K) = A(J,K) - A(J,I)*A(K,I)
			END DO	
		END DO
END DO
```
from [here](https://algowiki-project.org/en/Cholesky_decomposition). I am on it.

Reply:
So now when A is both a read_buffer and a write_buffer in the case of inplace cholesky, what do I classify it as?

Reply:
Probably best to add a new read_write_buffer type, even though write_buffer's semantics would work fine so far.

Reply:
We should probably add a read_write_buffer here right?

Reply:
Ah, Github playing on me by not refreshing the comments, but it seems that we agree :)